File: 098.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode * prev = new TreeNode(INT_MIN);  
    bool is = true; 
    bool prevAss = false; 
    bool isValidBST(TreeNode* root) {
        aux(root); 
        return is; 
    }
    void aux(TreeNode* root){
        if (!root) return; 
        aux(root->left); 
        if (root->val <= prev->val && prevAss){
            is = false;
            return; 
        }
        prev = root; 
        prevAss = true; 
        aux(root->right); 
    
    }
};

File: 098.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        ceil = float('inf')
        floor = float('-inf')
        return self.traverseTree(root.left,floor , root.val) and self.traverseTree(root.right, root.val, ceil) 
    
    
    def traverseTree(self, node, floor, ceil):
        if node == None: 
            return True 
        
        if node.val <= floor:
            return False 
        elif node.val >= ceil: 
            return False 
        
            
        isValidLeft = self.traverseTree(node.left, floor, min(ceil, node.val))
        isValidRight = self.traverseTree(node.right, max(floor, node.val), ceil)
        
        return isValidLeft and isValidRight
        

        
# Another similar code =======================================================================

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        less_than = float(inf)
        higher_than = float(-inf)
        
        return self.traverseTree(root, higher_than, less_than)
        
    def traverseTree(self, root, higher_than, less_than):
        if root == None:
            return True
        
        if root.val <= higher_than or root.val >= less_than:
            return False 
        
        
        isLeftAcceptable = self.traverseTree(root.left, higher_than, root.val)
        isRightAcceptable = self.traverseTree(root.right, root.val, less_than)
        
        return isLeftAcceptable and isRightAcceptable
        
        
